interface/src/app/signature/enriched_signatures.js

```diff
@@ -0,0 +1,203 @@
+  '$log', 'errGen', 'Participation', 'MathFuncts', 'Gene',
+  function EnrichedSignatureController($stateParams, Signature, $q, $log,
+    errGen, Participation, MathFuncts, Gene) {
+    var self = this;
+    self.isValidModel = false;
+    // Do nothing if the model ID in URL is falsey. The error will be taken
+    // care of by "<ml-model-validator>" component.
+    if (!$stateParams.mlmodel) {
+      return;
+    }
+
+    self.modelInUrl = $stateParams.mlmodel;
+    self.statusMessage = 'Connecting to the server ...';
+    self.pValueCutoff = 0.05;
+    self.enrichedSignatures = [];
+
+    // Do nothing if no genes are specified in URL.
+    if (!$stateParams.genes || !$stateParams.genes.split(',').length) {
+      self.statusMessage = 'No genes are specified.';
+      self.enrichedSignatures = [];
+      return;
+    }
+
+    var genesInUrl = [];
+    $stateParams.genes.split(',').forEach(function(token) {
+      var id = parseInt(token);
+      if (!isNaN(id) && genesInUrl.indexOf(id) === -1) {
+        genesInUrl.push(id);
+      }
+    });
+
+    // Promise that gets all signatures of given mlmodel in URL:
+    var p1 = Signature.get(
+      {mlmodel: self.modelInUrl, limit: 0},
+      function success(response) {
+        self.signatures = Object.create(null);
+        response.objects.forEach(function(element) {
+          self.signatures[element.id] = element.name;
+        });
+      },
+      function error(err) {
+        var message = errGen('Failed to get signatures: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    // Promise that gets the total number of genes:
+    var p2 = Gene.get(
+      {},
+      function success(response) {
+        self.geneNum = response.meta.total_count;
+      },
+      function error(err) {
+        var message = errGen('Failed to get total gene number: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var participations;
+    // Promise that gets all participation records that are related to
+    // the genes in URL:
+    var p3 = Participation.get(
+      {'related_genes': $stateParams.genes, 'limit': 0},
+      function success(response) {
+        participations = response.objects;
+      },
+      function error(err) {
+        var message = errGen(
+          'Failed to get signature-gene participations: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var pValueSigDigits = 3;
+    // groupedGenes is an object that behaves like a 2-D table of genes.
+    // groupedGenes[participationTypeID][signatureID] is an object whose
+    // keys are gene IDs and values are genes names of those genes that
+    // participate with signatureID and with participationTypeID.
+    var groupedGenes = Object.create(null);
```

Comment:
Thank you for the awesome comment explaining the structure of `groupedGenes` here! The variable isn't used until we get down to `getEnrichedSignatures()` at line 165 below. Could we move this comment and definition down there to minimize the scrolling needed to refer to this comment as a new reader tries to grok the code there?

Reply:
That's a good point. I will do that.